### PR TITLE
resource: enhance resource.drain RPC with "update" and "overwrite" modes

### LINKED ISSUE #4111: flux resource: require force option to drain a node that's already drained
Problem:  when automatically draining a node, if the node is already drained, the reason is overwritten.

Example: running `flux overlay disconnect` on a node that's running a job raises a "lost contact" fatal exception and drains the node, but the message associated with the drained state is from an error in the epilog.  Resource eventlog shows:
```
{"timestamp":1644271611.9434967,"name":"drain","context":{"idset":"100","reason":"broker was unresponsive"}}
{"timestamp":1644271699.4337404,"name":"offline","context":{"idset":"100"}}
{"timestamp":1644271699.4828186,"name":"drain","context":{"idset":"100","reason":"epilog failed for jobid ƒHtEVPW68d5"}}
```

We may want to add a "Force" flag to the RPC that drains nodes, and require the flag to be set to "re-drain" a node and update the reason.  The `flux admin drain` command could get a `--force` option also.

### PR DESCRIPTION
**Pull request by grondo**
This PR introduces some changes in behavior for the `resource.drain` RPC, reflected in a change in the default behavior of `flux resource drain` along with some additional options:

 * By default, a `resource.drain` RPC now fails if any targets are already drained. This matches the behavior of `flux resource undrain` and also avoids overwriting existing drain `reasons` without positive verification that you really mean it.
 * An optional `mode` member of the RPC request is introduced to change the default drain mode to either:
   - mode=`update`: The RPC does not fail if one or more targets are already drained, and does not overwrite existing `reason` for any drained target (it will, however, update a `reason` that is not set)
   - mode=`overwrite`: The RPC does not fail if one or more targets are already drained. All `reason` fields are overwritten
   - `flux resource drain` gets two new options `--force` (overwrite) and `--update` to optionally select one of these two modes.
  * The drain `timestamp` is only updated when the drain state changes. Thus the timestamp indicates when a node was drained, not when its reason field was last updated. (That timestamp can be obtained from the resource eventlog)
  * internally, the `resource` module uses "update" mode when draining ranks due to a resource topology issue such as missing resources. This preserves a reason possibly set by the instance owner/admin. The drain events due to topo errors can still be seen in the resource eventlog.
  * Internally, `reason=NULL` is now considered "unset". Places where `NULL` was replaced with `unknown` have been removed. An unset `reason` can be updated - it doesn't require `overwrite` mode.

Finally, the `flux perilog-run` command is switched to `update` mode so that valid `reason` fields are not overwritten when a rank is drained due to a failed prolog/epilog.

This PR might have some controversial choices, especially in how `NULL` reason is treated.  I was balancing most convenient with least surprising, and just decided to go with it. 

Fixes #4111

### CODE REVIEW

**Review by garlick — APPROVED** [review]
I tried this out and it all seems to work well.  This are good refinements that will definitely be needed on a real system IMHO - nice work anticipating them.

Just a few minor things to be addressed.

By the way, this was a model PR.  Each commit was focused and well documented, so walking through them was a super easy review.  Thanks for taking the time to do that!

File: src/modules/resource/drain.c [hunk 1]
```diff
@@ -119,6 +124,46 @@ static int update_draininfo_idset (struct drain *drain,
...
+    rank = idset_first (idset);
+    while (rank != IDSET_INVALID_ID) {
+        if (drain->info[rank].drained && drain->info[rank].reason) {
+            rc = -1;
+            (void) idset_set (errids, rank);
+        }
+        rank = idset_next (idset, rank);
+    }
+    if (rc < 0) {
+        char *s = idset_encode (errids, IDSET_FLAG_RANGE);
```
[the remark this answers is not in the document]
Comment (grondo) [reply src/modules/resource/drain.c:154]: This was done on purpose. If the `idset_encode()` fails then `"unknown"` is returned to the user, and I didn't really think at the time an extra log message was necessary. However, it doesn't hurt to add it so I'll go ahead and do that.

**Comment by grondo** [comment]
Thanks @garlick! I believe I've addressed all your comments and squashed the result (hope that was OK), and rebased.

**Comment by garlick** [comment]
LGTM, thanks!

**Comment by grondo** [comment]
Thanks!

**Pull request merged**